Components v2 (#7487)
hammerandchisel/discord-api-docs: docs/components/reference.mdx

---
sidebar_label: Component Reference
---

# Component Reference

This document serves as a comprehensive reference for all available components. It covers three main categories:

- **Layout Components** - For organizing and structuring content (Action Rows, Sections, Containers)
- **Content Components** - For displaying static text, images, and files (Text Display, Media Gallery, Thumbnails)
- **Interactive Components** - For user interactions (Buttons, Select Menus, Text Input)

To use these components, you need to send the [message flag](/docs/resources/message#message-object-message-flags) `1 << 15` (IS_COMPONENTS_V2) which can be sent on a per-message basis. This enables the new components system with the following changes:

> - The `content` and `embeds` fields are no longer available directly - they must be sent as components

Reviewer: See https://github.com/discord/discord-api-docs/pull/7487/files#r2053217633

- Attachments won't show by default - they must be exposed through components
- The `poll` and `stickers` fields are disabled

> > info

Reviewer 2: I will stop calling these out, but want to check for all of them

> > [Legacy component behavior](/docs/components/reference#legacy-message-component-behavior) will continue to work but provide less flexibility and control over the message layout.

Reviewer: ```suggestion
:::info
[Legacy component behavior](/docs/components/reference#legacy-message-component-behavior) will continue to work but provide less flexibility and control over the message layout.
:::
```

For a practical guide on implementing these components, see our [Using Message Components](/docs/components/using-message-components) and [Using Modal Components](/docs/components/using-modal-components) documentation.

---

## What is a Component

Components allow you to style and structure your messages, modals, and interactions. They are interactive elements that can create rich user experiences in your Discord applications.

> Components are a field on the [message object](/docs/resources/message#message-object). You can use them when creating messages or responding to an interaction, like an [application command](/docs/interactions/application-commands#).

Reviewer 3: or modals!

| 8    | [Channel Select](/docs/components/reference#channel-select)         | Select menu for channels                                   | Interactive | Message        |
| 9    | [Section](/docs/components/reference#section)                       | Container to display text alongside an accessory component | Layout      | Message        |
| 10   | [Text Display](/docs/components/reference#text-display)             | Markdown text                                              | Content     | Message        |
| 11   | [Thumbnail](/docs/components/reference#thumbnail)                   | Small image that can be used as an accessory               | Content     | Message        |
| 12   | [Media Gallery](/docs/components/reference#media-gallery)           | Display images and other media                             | Content     | Message        |
| 13   | [File](/docs/components/reference#file)                             | Displays an attached file                                  | Content     | Message        |
| 14   | [Separator](/docs/components/reference#separator)                   | Component to add vertical padding between other components | Layout      | Message        |

> | 17   | [Container](/docs/components/reference#container)                   | Container that visually groups a set of components         | Layout      | Message        |

Reviewer: I assume we decided against including the content inventory type (it can be received) as reference?

## Anatomy of a Component

All components have the following fields:

| Field | Type    | Description                                                                              |
|-------|---------|------------------------------------------------------------------------------------------|
| type  | integer | The [type](/docs/components/reference#component-object-component-types) of the component |
| id?   | integer | Optional identifier for component                                                        |

> The `id` field is optional and is used to identify components in the response from an interaction that aren't interactive components. The `id` must be unique within the message and is generated sequentially if left empty. Generation of `id`s won't use another `id` that exists in the message if you have one defined for another component.

Reviewer: Maybe specify the upper limit?

| emoji?    | partial [emoji](/docs/resources/emoji#emoji-object) | `name`, `id`, and `animated`                                                                                        |
| custom_id | string                                              | Developer-defined identifier for the button; max 100 characters                                                     |
| sku_id?   | snowflake                                           | Identifier for a purchasable [SKU](/docs/resources/sku#sku-object), only available when using premium-style buttons |
| url?      | string                                              | URL for link-style buttons                                                                                          |
| disabled? | boolean                                             | Whether the button is disabled (defaults to `false`)                                                                |

Buttons come in various styles to convey different types of actions. These styles also define what fields are valid for a button.

> - Non-link and Non-premium buttons **must** have a `custom_id`, and cannot have a `url` or a `sku_id`.

Reviewer: ```suggestion
- Non-link and non-premium buttons **must** have a `custom_id`, and cannot have a `url` or a `sku_id`.
```
Unsure but lowercase n because it's not the beginning of the sentence?

Author: Good catch!

        ],
        "channel_id": "345626669114982402",
        "author": {
            "username": "Mason",
            "public_flags": 131141,
            "id": "53908232506183680",
            "discriminator": "1337",
            "avatar": "a_d5efa99b3eeaa7dd43acca82f5692432"
        },
        "attachments": []
    },
    "member": {
        "user": {
            "username": "Mason",
            "public_flags": 131141,
            "id": "53908232506183680",
            "discriminator": "1337",
            "avatar": "a_d5efa99b3eeaa7dd43acca82f5692432"
        },
        "roles": [
            "290926798626357999"
        ],
        "premium_since": null,
        "permissions": "17179869183",
        "pending": false,
        "nick": null,
        "mute": false,
        "joined_at": "2017-03-13T19:19:14.040000+00:00",
        "is_pending": false,
        "deaf": false,
        "avatar": null
    },
    "id": "846462639134605312",
    "guild_id": "290926798626357999",
    "data": {
        "custom_id": "click_one",
        "component_type": 2
    },
    "channel_id": "345626669114982999",
    "application_id": "290926444748734465"
}
```

### Button Design Guidelines

> ###### General Button copy

Reviewer 4: ```suggestion
###### General Button Guidelines
```
Or something like that, not sure how the "copy" got in there and if that's intentional

Author: The term "copy" in this context refers to the text used in advertising, marketing, or published materials. I don't think it's as common as I thought it was though so I'll swap this (and make sure it's capitalized!)

                ],
            }
        ],
        "edited_timestamp": null,
        "flags": 32768,
        "id": "847587334500646933",
        "interaction": {
            "id": "847587333942935632",
            "name": "dropdown",
            "type": 2,
            "user": {
                "avatar": "a_d5efa99b3eeaa7dd43acca82f5692432",
                "discriminator": "1337",
                "id": "53908232506183680",
                "public_flags": 131141,
                "username": "Mason"
            }
        },
        "mention_everyone": false,
        "mention_roles":[],
        "mentions":[],
        "pinned": false,
        "timestamp": "2021-05-27T21:29:27.956000+00:00",
        "tts": false,
        "type": 20,
        "webhook_id": "845027738276462632"
    },
    "token": "UNIQUE_TOKEN",
    "type": 3,
    "version": 1
}
```

###### Select Menu Resolved Object

The `resolved` object is included in interaction payloads for user, role, mentionable, and channel select menu components. `resolved` contains a nested object with additional details about the selected options with the key of the resource type—`users`, `roles`, `channels`, and `members`.

> > info
> > `members` and `users` may both be present in the `resolved` object when a user is selected (in either a [user select](/docs/components/reference#user-select) or [mentionable select](/docs/components/reference#mentionable-select)).

Reviewer: ```suggestion
:::info
`members` and `users` may both be present in the `resolved` object when a user is selected (in either a [user select](/docs/components/reference#user-select) or [mentionable select](/docs/components/reference#mentionable-select)).
:::
```

| disabled?       | boolean                                                                                                 | Whether select menu is disabled (defaults to `false`)                                                                                                    |

###### Example

```json
"components": [
  {
    "type": 1,
    "components": [
      {
        "type": 8,
        "custom_id": "channel_select",
        "channel_types": [0],
        "placeholder": "Which text channel?"
      }
    ]
  }
]
```

![Example of a Channel Select for text channels](images/components/channel-select.png)

###### Channel Select Interaction

Check out the [String Select interaction](/docs/components/reference#string-select-string-select-interaction) to see the structure of the Select interaction object and resolved object.

## Section

A Section is a top-level layout component that allows you to join text contextually with an accessory.

Sections are only available in messages.

> > info
> > To use this component, you need to send the [message flag](/docs/resources/message#message-object-message-flags) `1 << 15` (IS_COMPONENTS_V2) which can be activated on a per-message basis.

Reviewer: ```suggestion
:::info
To use this component, you need to send the [message flag](/docs/resources/message#message-object-message-flags) `1 << 15` (IS_COMPONENTS_V2) which can be activated on a per-message basis.
:::
```

|------------|-----------------------------------------------------------------------------------------------------------------------|---------------------------------------------------------------------------------------------------|
| type       | integer                                                                                                               | `9` for section component                                                                         |
| id?        | integer                                                                                                               | Optional identifier for component                                                                 |
| components | array of [text display components](/docs/components/reference#text-display)                                           | One to three text components                                                                      |
| accessory  | [thumbnail component](/docs/components/reference#thumbnail) or [button components](/docs/components/reference#button) | A thumbnail or a button component, with a future possibility of adding more compatible components |

> > info
> > Don't hardcode `components` to contain only text components. We may add other components in the future. Similarly, `accessory` may be expanded to include other components in the future.

Reviewer: ```suggestion
:::info
Don't hardcode `components` to contain only text components. We may add other components in the future. Similarly, `accessory` may be expanded to include other components in the future.
:::
```

###### Example

```json
"components": [
  {
    "type": 9,
    "components": [
      {
        "type": 10,
        "content": "# Real Game v7.3"
      },
      {
        "type": 10,
        "content": "Hope you're excited, the update is finally here! Here are some of the changes:\n- Fixed a bug where certain treasure chests wouldn't open properly\n- Improved server stability during peak hours\n- Added a new type of gravity that will randomly apply when the moon is visible in-game\n- Every third thursday the furniture will scream your darkest secrets to nearby npcs"
      },
      {
        "type": 10,
        "content": "-# That last one wasn't real, but don't use voice chat near furniture just in case..."
      }
    ],
    "accessory": {
      "type": 11,
      "media": {
        "url": "websitewithopensourceimages/gamepreview.png"
      }
    }
  }
]
```

> ![Example of a Section showing a fake game chagnelog and a thumbnail](images/components/section.png)

Reviewer 4: ```suggestion
![Example of a Section showing a fake game changelog and a thumbnail](images/components/section.png)
```

## Text Display

A Text Display is a top-level content component that allows you to add text to your message formatted with markdown. This is similar to the `content` field of a message, but allows you to add multiple text components, controlling the layout of your message.

Text Displays are only available in messages.

> > info
> > To use this component, you need to send the [message flag](/docs/resources/message#message-object-message-flags) `1 << 15` (IS_COMPONENTS_V2) which can be activated on a per-message basis.

Reviewer: ```suggestion
:::info
To use this component, you need to send the [message flag](/docs/resources/message#message-object-message-flags) `1 << 15` (`IS_COMPONENTS_V2`) which can be activated on a per-message basis.
:::
```

| Field   | Type    | Description                                      |
|---------|---------|--------------------------------------------------|
| type    | integer | `10` for text display                            |
| id?     | integer | Optional identifier for component                |
| content | string  | Text that will be displayed similar to a message |

###### Example

```json
"components": [
  {
    "type": 10,
    "content": "# This is a Text Display\nAll the regular markdown rules apply\n- You can make lists\n- You can use `code blocks`\n- You can use [links](https://discord.com/)\n- Even :blush: :star_struck: :exploding_head:\n- Spolier alert: ||these too!||"
  }
]
```

![Example of a Text Display with markdown](images/components/text-display.png)

## Thumbnail

A Thumbnail is a content component that is a small image only usable as an accessory in a [section](/docs/components/reference#section). The preview comes from an url or attachment through the [unfurled media item](/docs/components/reference#unfurled-media-item-structure) structure.

Thumbnails are only available in messages as an accessory in a [section](/docs/components/reference#section).

> > info
> > To use this component, you need to send the [message flag](/docs/resources/message#message-object-message-flags) `1 << 15` (IS_COMPONENTS_V2), which can be activated on a per-message basis.

Reviewer: ```suggestion
:::info
To use this component, you need to send the [message flag](/docs/resources/message#message-object-message-flags) `1 << 15` (`IS_COMPONENTS_V2`), which can be activated on a per-message basis.
:::
```

| id?          | integer                                                                         | Optional identifier for component                                               |
| media        | [unfurled media item](/docs/components/reference#unfurled-media-item-structure) | A url or attachment                                                             |
| description? | string                                                                          | Alt text for the media                                                          |
| spoiler?     | boolean                                                                         | Whether the thumbnail should be a spoiler (or blurred out). Defaults to `false` |

###### Example

Check out the [Section example](/docs/components/reference#section-example) to see the thumbnail in use as an accessory.

## Media Gallery

A Media Gallery is a top-level content component that allows you to display 1-10 media attachments in an organized gallery format. Each item can have optional descriptions and can be marked as spoilers.

Media Galleries are only available in messages.

> > info
> > To use this component, you need to send the [message flag](/docs/resources/message#message-object-message-flags) `1 << 15` (IS_COMPONENTS_V2) which can be activated on a per-message basis.

Reviewer: ```suggestion
:::info
To use this component, you need to send the [message flag](/docs/resources/message#message-object-message-flags) `1 << 15` (`IS_COMPONENTS_V2`) which can be activated on a per-message basis.
:::
```

| spoiler?     | boolean                                                                         | Whether the media should be a spoiler (or blurred out). Defaults to `false` |

###### Example

```json
"components": [
  {
    "type": 10,
    "content": "Live webcam shots as of 18-04-2025 at 12:00 UTC"
  },
  {
    "type": 12,
    "items": [
      {
        "media": {"url": "livevideofeedconvertedtoimage/webcam1.png"},
        "description": "Feed 1"
      },
      {
        "media": {"url": "livevideofeedconvertedtoimage/webcam2.png"},
        "description": "Feed 2"
      },
      {
        "media": {"url": "livevideofeedconvertedtoimage/webcam3.png"},
        "description": "Feed 3"
      }
    ]
  }
]
```

![Example of a Media Gallery showing screenshots from live webcam feeds](images/components/media-gallery.png)

## File

A File is a top-level component that allows you to display an uploaded file as an attachment to the message and reference it in the component. This is similar to the `embeds` field of a message but allows you to control the layout of your message by using this anywhere as a component. Files are only available in messages.

> > info
> > To use this component, you need to send the [message flag](/docs/resources/message#message-object-message-flags) `1 << 15` (IS_COMPONENTS_V2) which can be activated on a per-message basis.

Reviewer: ```suggestion
:::info
To use this component, you need to send the [message flag](/docs/resources/message#message-object-message-flags) `1 << 15` (`IS_COMPONENTS_V2`) which can be activated on a per-message basis.
:::
```
```

| file     | [unfurled media item](/docs/components/reference#unfurled-media-item-structure) | This unfurled media item is unique in that it **only** support attachment references using the `attachment://<filename>` syntax |
| spoiler? | boolean                                                                         | Whether the media should be a spoiler (or blurred out). Defaults to `false`                                                     |

###### Example

```json
"components": [
  {
    "type": 10,
    "content": "# New game version released for testing!"
  },
  {
    "type": 10,
    "content": "Grab the game here:"
  },
  {
    "type": 13,
    "file": {
      "url": "attachment://game.zip"
    }
  },
  {
    "type": 10,
    "content": "Latest manual artwork here:"
  },
  {
    "type": 13,
    "file": {
      "url": "attachment://manual.pdf"
    }
  }
]
```

![Example of a File showing a download for a game and manual](images/components/file.png)

## Separator

A Separator is a top-level layout component that adds vertical padding and visual division between other components.

Separators are only available in messages.

> > info
> > To use this component, you need to send the [message flag](/docs/resources/message#message-object-message-flags) `1 << 15` (IS_COMPONENTS_V2) which can be activated on a per-message basis.

Reviewer: ```suggestion
:::info
To use this component, you need to send the [message flag](/docs/resources/message#message-object-message-flags) `1 << 15` (`IS_COMPONENTS_V2`) which can be activated on a per-message basis.
:::
```

|----------|---------|-----------------------------------------------------------------------------------------|
| type     | integer | `14` for separator component                                                            |
| id?      | integer | Optional identifier for component                                                       |
| divider? | boolean | Whether a visual divider should be displayed in the component. Defaults to `true`       |
| spacing? | integer | Size of separator padding—`1` for small padding, `2` for large padding. Defaults to `1` |

###### Example

```json
"components": [
  {
    "type": 10,
    "content": "Above"
  },
  {
    "type": 14,
    "divider": true,
    "spacing": 2
  },
  {
    "type": 10,
    "content": "Below"
  }
]
```

![Example of a separator with large spacing dividing content](images/components/separator.png)

## Container

A Container is a top-level layout component that holds up to 10 components. Containers are visually distinct from surrounding components and has an optional customizable color bar.

Containers are only available in messages.

> > info
> > To use this component, you need to send the [message flag](/docs/resources/message#message-object-message-flags) `1 << 15` (IS_COMPONENTS_V2) which can be activated on a per-message basis.

Reviewer: ```suggestion
:::info
To use this component, you need to send the [message flag](/docs/resources/message#message-object-message-flags) `1 << 15` (`IS_COMPONENTS_V2`) which can be activated on a per-message basis.
:::
```

|---------------|----------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| type          | integer                    | `17` for container component                                                                                                                                                                                                                                                                                                                      |
| id?           | integer                    | Optional identifier for component                                                                                                                                                                                                                                                                                                                 |

> | components    | array of select components | Up to 10 components of type [action row](/docs/components/reference#action-row), [text display](/docs/components/reference#text-display), [section](/docs/components/reference#section), [media gallery](/docs/components/reference#media-gallery), [separator](/docs/components/reference#separator), or [file](/docs/components/reference#file) |

Reviewer 3: "array of select components" could be confused with the "Select" component types

Reviewer: ```suggestion
| components    | array of components | Up to 10 components of the type [action row](/docs/components/reference#action-row), [text display](/docs/components/reference#text-display), [section](/docs/components/reference#section), [media gallery](/docs/components/reference#media-gallery), [separator](/docs/components/reference#separator), or [file](/docs/components/reference#file) |
```
(This will break the table)

Author: Will update this one, thanks!

      {
        "type": 10,
        "content": "# You have encountered a wild coyote!"
      },
      {
        "type": 12,
        "items": [
          {
            "media": {"url": "websitewithopensourceimages/coyote.png"},
          }
        ]
      },
      {
        "type": 10,
        "content": "What would you like to do?"
      },
      {
        "type": 1,
        "components": [
          {
            "type": 2,
            "custom_id": "pet_coyote",
            "label": "Pet it!",
            "style": 1
          },
          {
            "type": 2,
            "custom_id": "feed_coyote",
            "label": "Attempt to feed it",
            "style": 2
          },
          {
            "type": 2,
            "custom_id": "run_away",
            "label": "Run away!",
            "style": 4
          }
        ]
      }
    ]
  }
]
```

![Example of a container showing text, image, and buttons for a wild enemy encounter](images/components/container.png)

## Unfurled Media Item Structure

| Field | Type   | Description                                                     |
|-------|--------|-----------------------------------------------------------------|

> | url   | string | Supports arbitrary urls and attachment://\<filename> references |

Reviewer 5: When receiving the component, this will have extra fields like `proxy_url`, `height`, `width`, `content_type` and `loading_state`. Will they be documented too?

Reviewer: If possible Anthony, please document the resolved media object. 
We had this discussion a few times with dv8.
See https://canary.discord.com/channels/1317206872763404478/1347953940263731263 as reference.

Author: Documented, thanks for bringing it up!

Reviewer 5: @anthonydiscord Sorry for bothering, but wanted to clarify something.

This payload was found in the wild a while ago which has most fields as null, is that something we would expect or behaviour that will no longer be seen:
```json
{
  "media": {
    "url": "<readacted>",
    "proxy_url": "<readacted>",
    "width": null,
    "height": null,
    "placeholder": null,
    "placeholder_version": null,
    "content_type": null,
    "loading_state": 3,
    "flags": 0
  },
  "description": null,
  "spoiler": false
}
```

:::info
Don't hardcode `components` to contain only text components. We may add other components in the future. Similarly, `accessory` may be expanded to include other components in the future.
:::

###### Example

```json
"components": [
  {
    "type": 9,
    "components": [
      {
        "type": 10,
        "content": "# Real Game v7.3"
      },
      {
        "type": 10,
        "content": "Hope you're excited, the update is finally here! Here are some of the changes:\n- Fixed a bug where certain treasure chests wouldn't open properly\n- Improved server stability during peak hours\n- Added a new type of gravity that will randomly apply when the moon is visible in-game\n- Every third thursday the furniture will scream your darkest secrets to nearby npcs"
      },
      {
        "type": 10,
        "content": "-# That last one wasn't real, but don't use voice chat near furniture just in case..."
      }
    ],
    "accessory": {
      "type": 11,
      "media": {

>         "url": "websitewithopensourceimages/gamepreview.png"

Reviewer 6: ```suggestion
        "url": "https://websitewithopensourceimages/gamepreview.png"
```

###### Media Gallery Item Structure

| Field        | Type                                                                            | Description                                                                 |
|--------------|---------------------------------------------------------------------------------|-----------------------------------------------------------------------------|
| media        | [unfurled media item](/docs/components/reference#unfurled-media-item-structure) | A url or attachment                                                         |
| description? | string                                                                          | Alt text for the media                                                      |
| spoiler?     | boolean                                                                         | Whether the media should be a spoiler (or blurred out). Defaults to `false` |

###### Example

```json
"components": [
  {
    "type": 10,
    "content": "Live webcam shots as of 18-04-2025 at 12:00 UTC"
  },
  {
    "type": 12,

>     "items": [
>       {
>         "media": {"url": "livevideofeedconvertedtoimage/webcam1.png"},
>         "description": "Feed 1"
>       },
>       {
>         "media": {"url": "livevideofeedconvertedtoimage/webcam2.png"},
>         "description": "Feed 2"
>       },
>       {
>         "media": {"url": "livevideofeedconvertedtoimage/webcam3.png"},
>         "description": "Feed 3"
>       }
>     ]

Reviewer 6: ```suggestion
    "items": [
      {
        "media": {"url": "https://livevideofeedconvertedtoimage/webcam1.png"},        
        "description": "An aerial view looking down on older industrial complex buildings. The main building is white with many windows and pipes running up the walls."
      },
      {
        "media": {"url": "https://livevideofeedconvertedtoimage/webcam2.png"},
        "description": "An aerial view of old broken buildings. Nature has begun to take root in the rooftops. A portion of the middle building's roof has collapsed inward. In the distant haze you can make out a far away city."
      },
      {
        "media": {"url": "attachment://webcam3.png"},
        "description": "A street view of a downtown city. Prominently in photo are skyscrapers and a domed building"
      }
    ]
```

|---------------|---------|--------------------------------------------------------------------------------------------------------------------------------------------------|
| url           | string  | Supports arbitrary urls and attachment://\<filename> references                                                                                  |
| proxy_url?    | string  | The proxied url of the media item. This field is ignored and provided by the API as part of the response                                         |
| height?       | integer | The height of the media item. This field is ignored and provided by the API as part of the response                                              |
| width?        | integer | The width of the media item. This field is ignored and provided by the API as part of the response                                               |
| content_type? | string  | The [media type](https://en.wikipedia.org/wiki/Media_type) of the content. This field is ignored and provided by the API as part of the response |

### Uploading a file

> You can upload a file when sending a message and reference it within a file component using the `attachment://<filename>` syntax. Each file component can only display 1 attached file, but you can upload multiple files and add them to different file components within your payload.

Reviewer 6: This shouldn't reference the `file` component. Perhaps 
```suggestion
You can upload a file when sending a message and reference it within an Unfurled Media Item using the `attachment://<filename>` syntax in the `url` field. 
```

Author: Oh yeah this was initially in the file section and I missed that when I moved it, good call thank you

To upload a file with your message, you'll need to send your payload as `multipart/form-data` (rather than `application/json`) and include your file with a valid filename in your payload. Details and examples for uploading files can be found in the [API Reference](/docs/reference#uploading-files).

## Legacy Message Component Behavior

> Before the introduction of the `IS_COMPONENTS_V2` flag ([see changelog](/docs/change-log/2025-04-22-components-v2)), message components were sent as part of the message content. This means that you could send a message with components without setting the `IS_COMPONENTS_V2` flag, and the components would be included in the message content along with `content` and `embeds`.

Reviewer 6: ```suggestion
Before the introduction of the `IS_COMPONENTS_V2` flag ([see changelog](/docs/change-log/2025-04-22-components-v2)), message components were sent in conjunction with message content. This means that you could send a message using a subset of the available components without setting the `IS_COMPONENTS_V2` flag, and the components would be included in the message content along with `content` and `embeds`.
```
